gnome-image-installer/util/eos-reformatter.c

```diff
@@ -0,0 +1,1088 @@
+            }
+        }
+
+      g_async_queue_push (outq, buf);
+    }
+
+  return ret;
+}
+
+static gboolean
+eos_reformatter_decompress (EosReformatter *reformatter, GAsyncQueue *freeq, GAsyncQueue *inq, GAsyncQueue *outq)
+{
+  GConverterResult res;
+  GConverterFlags flags = G_CONVERTER_NO_FLAGS;
+  GError *error = NULL;
+  gsize input_read = 0;
+  EosAlignedBuffer *buf;
+  EosAlignedBuffer *outbuf;
+
+  if (g_cancellable_is_cancelled (reformatter->cancellable))
+    {
+      return FALSE;
+    }
+
+  buf = g_async_queue_pop (inq);
+  outbuf = eos_reformatter_get_free_buffer (reformatter, freeq);
+
+  if (buf->len == 0)
+    flags |= G_CONVERTER_INPUT_AT_END;
+
+  outbuf->len = 0;
+
+  /* Loop over input until consumed, sending possible output immediately */
+  do
+    {
+      gsize in = 0;
+      res = g_converter_convert (reformatter->decompressor,
+                                 buf->ptr + input_read, buf->len - input_read,
+                                 outbuf->ptr, reformatter->buffer_size,
+                                 flags, &in, &outbuf->len,
+                                 &error);
+      input_read += in;
+
+      if (outbuf->len > 0)
+        {
+          g_async_queue_push (outq, outbuf);
+          /* Note that this competes with read buffers if in different threads */
+          outbuf = eos_reformatter_get_free_buffer (reformatter, freeq);
+          if (g_cancellable_is_cancelled (reformatter->cancellable))
+            {
+              res = G_CONVERTER_ERROR;
+            }
+        }
+    }
+  while (res == G_CONVERTER_CONVERTED && input_read < buf->len);
+
+  /* Return the outbuf to free queue if we didn't use it */
+  if (outbuf->len == 0)
+    {
+      g_async_queue_push (freeq, outbuf);
+    }
+
+  /* This is fine, we just need to cycle to get more input buffers */
+  if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_PARTIAL_INPUT))
+    {
+      res = G_CONVERTER_CONVERTED;
+      g_error_free (error);
+      error = NULL;
+    }
+  else if (error != NULL)
+    {
+      g_mutex_lock (&reformatter->thread_mutex);
+      reformatter->error = error;
+      g_mutex_unlock (&reformatter->thread_mutex);
+      g_cancellable_cancel (reformatter->cancellable);
+    }
+
+  if (res == G_CONVERTER_FINISHED)
+    {
+      /* If conversion is finished, we send an empty buffer to indicate EOS */
+      buf->len = 0;
+      g_async_queue_push (outq, buf);
+    }
+  else
+    {
+      g_async_queue_push (freeq, buf);
+    }
+
+  return (res == G_CONVERTER_CONVERTED);
+}
+
+static gint
+eos_reformatter_prepare_write (EosReformatter *reformatter)
+{
+  if (reformatter->write_fd > 0)
+    return reformatter->write_fd;
+
+  return open(reformatter->device, O_WRONLY | O_SYNC | O_CREAT);
+}
+
+static gboolean
+eos_reformatter_add_sample (gpointer data)
+{
+  EosReformatter *reformatter = EOS_REFORMATTER(data);
+
+  gdu_estimator_add_sample (reformatter->estimator, reformatter->write_completed_bytes);
```

Comment:
`write_completed_bytes` is modified by a different thread so this read should be guarded by the mutex too, no?